fedimint-server/src/lib.rs

```diff
@@ -212,40 +212,61 @@ impl FedimintServer {
-                    last_outcome.epoch,
-                    contributions,
-                    None,
-                    self.last_processed_epoch.as_ref(),
-                )
-            } else {
-                info!("Downloading missing epoch {}", epoch_num);
-                let epoch_pk = self.cfg.epoch_pk_set.public_key();
-                self.api
-                    .fetch_epoch_history(epoch_num, epoch_pk)
-                    .await
-                    .expect("fetches history")
-            };
-
-            // We can't validate a hash on a tx we have just created ourselves and that doesn't
-            // have a `rejected_tx` yet
-            if current_epoch.outcome.rejected_txs.is_some() {
-                current_epoch.verify_hash(&self.last_processed_epoch)?;
-            }
-            epochs.push(current_epoch.clone());
-
-            let pk = self.cfg.epoch_pk_set.public_key();
-            if epoch_num == last_outcome.epoch || current_epoch.verify_sig(&pk).is_ok() {
-                for epoch in epochs.drain(..) {
-                    let outcome = ConsensusOutcomeConversion::from(epoch.outcome.clone()).0;
-                    self.consensus
-                        .process_consensus_outcome(outcome, &epoch.outcome.rejected_txs)
+            let (items, epoch, prev_epoch_hash, rejected_txs, at_know_trusted_checkpoint) =
+                if epoch_num == last_outcome.epoch {
+                    (
+                        last_outcome
+                            .contributions
+                            .iter()
+                            .sorted_by_key(|(peer, _)| *peer)
+                            .map(|(peer, items)| (*peer, items.clone()))
+                            .collect(),
+                        last_outcome.epoch,
+                        self.last_processed_epoch
+                            .as_ref()
+                            .map(|epoch| epoch.outcome.hash()),
+                        None,
+                        true,
+                    )
+                } else {
+                    info!("Downloading missing epoch {}", epoch_num);
+                    let epoch_pk = self.cfg.epoch_pk_set.public_key();
+                    let epoch = self
+                        .api
+                        .fetch_epoch_history(epoch_num, epoch_pk)
+                        .await
+                        .expect("fetches history");
+
+                    epoch.verify_hash(&self.last_processed_epoch)?;
+
+                    let pk = self.cfg.epoch_pk_set.public_key();
+                    let sig_valid = epoch.verify_sig(&pk).is_ok();
+                    (
+                        epoch.outcome.items,
+                        epoch.outcome.epoch,
+                        epoch.outcome.last_hash,
+                        epoch.outcome.rejected_txs,
+                        sig_valid,
+                    )
+                };
+
+            epochs.push((items, epoch, prev_epoch_hash, rejected_txs));
```

Comment:
Do we even need `prev_epoch_hash` rn? We should keep it though since it's probably needed to fix #1116.